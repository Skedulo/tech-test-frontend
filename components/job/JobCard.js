import React, { Component } from 'react'
import moment from 'moment'
const TimeFormat = "HH:mm";

export class JobCard extends Component {
    render() {
        return (
            <div className="job-card p-2 d-flex flex-column mb-3">
              <div className="p-2">
                <div className="name"><span className="font-weight-bold">{this.props.job.name}</span> <span className="job-order">(Job #{this.props.index})</span></div>
                <div className="location mb-2 secondary-info">{this.props.job.location}</div>
                <div className="date secondary-info">{new Date(this.props.job.start).toDateString()}</div>
                <div className="time font-weight-bold secondary-info">{moment(this.props.job.start).format(TimeFormat)} - {moment(this.props.job.end).format(TimeFormat)}</div>
              </div>
            </div>
        )
    }
}