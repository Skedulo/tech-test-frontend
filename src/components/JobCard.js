/** @jsx jsx */
import { jsx } from "@emotion/core";
import styled from "@emotion/styled";
import Card from "./Card";
import { Title, Subtitle, BodySpan } from "./Text";

// TODO: Use a library to format dates reliably
const formatDate = dateString => new Date(dateString).toDateString();
const formatTime = dateString => new Date(dateString).toLocaleTimeString();

const FlexColumn = styled.div`
  display: flex;
  flex-direction: column;
`;

const FlexRow = styled.div(
  ({ theme }) => `
display: flex;
justify-content: flex-start;
align-items: baseline;
`
);

const CardTitle = styled(Title)(
  ({ theme }) => `
margin-right: ${theme.spacing.sm};
margin-bottom: 0;
`
);
const CardSubtitle = styled(Subtitle)(
  ({ theme }) => `
margin-bottom: 0;
flex-shrink: 0;
`
);

const CardDetail = styled(BodySpan)(
  ({ theme }) => `
  margin-bottom: ${theme.spacing.unit * 2}px;
  color: ${theme.background.contrastMedium};
`
);

const DateText = styled(BodySpan)(
  ({ theme }) => `
  color: ${theme.background.contrastMedium};
`
);

// TODO: Handle jobs that start and finish on different days
const JobDate = ({ start, end }) => {
  return (
    <FlexColumn>
      <DateText>{formatDate(start)}</DateText>
      <DateText>{`${formatTime(start)} - ${formatTime(end)}`}</DateText>
    </FlexColumn>
  );
};

const JobCard = ({ id, name, start, end, cardDetail, className }) => (
  <Card className={className}>
    <FlexColumn>
      <FlexRow>
        <CardTitle>{name}</CardTitle>
        <CardSubtitle>(JOB-{id})</CardSubtitle>
      </FlexRow>
      <CardDetail>{cardDetail}</CardDetail>
      <JobDate start={start} end={end} />
    </FlexColumn>
  </Card>
);

export default JobCard;
