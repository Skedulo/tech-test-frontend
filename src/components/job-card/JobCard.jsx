import React from 'react';
import './styles.css';
import moment from 'moment'
import 'moment-timezone';

export const DATE_FORMAT = 'MMMM Do YYYY, h:mm:ss a';

export default React.memo(props => {
    const { name, start, end, contact } = props;
    const startTime = moment(start).format(DATE_FORMAT);
    const endTime = moment(end).format(DATE_FORMAT);

    return (
        <div className="job-card__container">
            <p className="job-card__name"><span>Job</span>{name}</p>
            <p className="job-card__start-time"><span>Start Time</span>{startTime}</p>
            <p className="job-card__end-time"><span>End Time</span>{endTime}</p>
            <p className="job-card__contact-name"><span>Contact</span>{contact.name}</p>
        </div>
    );
});
