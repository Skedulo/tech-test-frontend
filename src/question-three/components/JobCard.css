.jobCard {
    --sidebar-item-size: 2rem;
    background: var(--white);
    padding: var(--spacing--xs);
    display: grid;
    grid-template:
        [row1-start] "title title" auto [row1-end]
        [row2-start] "location location" auto [row2-end]
        [row2-start] "datetime allocationCounter" auto [row2-end]
          / 1fr auto;
}

.jobCard__title {
    grid-area: title;
}

.jobCard__title__id {
    color: var(--gray);
    font-size: var(--font-size--s);
}

.jobCard__location  {
    grid-area: location;
    color: var(--gray);
    font-size: var(--font-size--xs);
}

.jobCard__datetime  {
    grid-area: datetime;
    font-size: var(--font-size--xs);
    display: grid;
    grid-auto-flow: row;
    align-content: end;
}

.jobCard__datetime__date  {
    color: var(--gray);
}
.jobCard__datetime__time  {
    color: var(--darkerGray);
}

.jobCard__allocationCounter {
    grid-area: allocationCounter;
    border-radius: 100%;
    border: 1px solid var(--primary);
    color: var(--primary);
    width: var(--sidebar-item-size);
    height: var(--sidebar-item-size);
    display: grid;
    justify-content: center;
    align-content: center;
}


.jobCard__allocationCounter--hidden {
    visibility: hidden;
}