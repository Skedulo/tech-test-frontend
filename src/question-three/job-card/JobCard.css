.question-three__job-card {
  background: #FFFFFF;
  padding: 10px;
  margin: 15px;
}

.question-three__secondary-text {
  color: #C4C4C4;
}

.question-three__job-date-time {
  color: #8D8383;
}

.question-three__job-id {
  font-size: 16px;
}

.question-three__job-name {
  font-size: 20px;
}