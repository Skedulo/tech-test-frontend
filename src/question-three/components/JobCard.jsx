import React from 'react'
import PropTypes from 'prop-types'

import './JobCard.css'
const formatTime = (date) => `${date.getUTCHours().toString().padStart(2, '0')}:${date.getUTCMinutes().toString().padStart(2, '0')}`

const JobCard = ({ job: { name, location, id, start, end, jobAllocations } }) => {
  const startDate = new Date(start)
  const endDate = new Date(end)
  return (
    <div className="jobCard">
      <div className="jobCard__title">
        <strong className="jobCard__title__name">{name}</strong>
        <span className="jobCard__title__id"> (Job #{id})</span>
      </div>
      <div className="jobCard__location">{location}</div>
      <div className="jobCard__datetime">
        <div className="jobCard__datetime__date">{startDate.toDateString()}</div>
        <div className="jobCard__datetime__time">
          {formatTime(startDate)}-{formatTime(endDate)}
        </div>
      </div>
      <div className={`jobCard__allocationCounter ${!jobAllocations.length ? 'jobCard__allocationCounter--hidden' : ''}`}>
        {jobAllocations.length}
      </div>
    </div>
  )
}

JobCard.propTypes = {
  job: PropTypes.shape({
    name: PropTypes.string,
    location: PropTypes.string,
    id: PropTypes.string,
    start: PropTypes.instanceOf(Date),
    end: PropTypes.instanceOf(Date),
    jobAllocations: PropTypes.array
  })
}

export default JobCard
