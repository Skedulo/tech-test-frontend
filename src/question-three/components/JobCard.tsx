import React from 'react'
import { Job } from '../../types/Job'

import './JobCard.css'
const formatTime = (date: Date) => `${date.getUTCHours().toString().padStart(2, '0')}:${date.getUTCMinutes().toString().padStart(2, '0')}`

const JobCard: React.FC<{job: Job}> = ({ job: {start, end, name, id, location, jobAllocations = []} }) => {
  const startDate = new Date(start)
  const endDate = new Date(end)
  return (
    <div className="jobCard">
      <div className="jobCard__title">
        <strong className="jobCard__title__name">{name}</strong>
        <span className="jobCard__title__id"> (Job #{id})</span>
      </div>
      <div className="jobCard__location">{location}</div>
      <div className="jobCard__datetime">
        <div className="jobCard__datetime__date">{startDate.toDateString()}</div>
        <div className="jobCard__datetime__time">
          {formatTime(startDate)}-{formatTime(endDate)}
        </div>
      </div>
      <div className={`jobCard__allocationCounter ${!jobAllocations.length ? 'jobCard__allocationCounter--hidden' : ''}`}>
        {jobAllocations.length}
      </div>
    </div>
  )
}

export default JobCard
