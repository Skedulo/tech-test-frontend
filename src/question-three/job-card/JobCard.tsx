import React from "react"
import "./JobCard.css"
import { Job } from "../../common/types"
import { formatDate, formatTime } from "../../common/formatDateTime";

interface IJobItem {
  job: Job,
}

export const JobCard: React.FC<IJobItem> = (props: IJobItem) => {
  const { job } = props;
  return (
    <div className="question-three__job-card" key={job.id}>
      <div className="question-three__job-name">
        {job.name} <span className="question-three__job-id question-three__secondary-text">(Job #{job.id})</span>
      </div>
      <div className="question-three__job-location question-three__secondary-text">
        {job.location}
      </div>
      <div className="question-three__job-date question-three__secondary-text">
        {formatDate(job.start)}
      </div>
      <div className="question-three__job-date-time question-three__secondary-text">
        {formatTime(job.start)} - {formatTime(job.end)}
      </div>
    </div>
  )
}

