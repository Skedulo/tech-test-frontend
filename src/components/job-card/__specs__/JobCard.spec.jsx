import React from 'react';
import { shallow } from 'enzyme';
import JobCard, { DATE_FORMAT } from '../JobCard';
import moment from 'moment';

describe('JobCard Component', () => {

  describe('snapshots', () => {
      const defaultProps = {
          name: 'Jean-Ralphio Saperstein',
          start: '2018-09-01T15:00:00Z',
          end: '2018-09-01T16:00:00Z',
          contact: { name: 'Tom Haverford'}
      };

    const renderComponent = (propOverrides = {}) => {
      const props = {
        ...defaultProps,
        ...propOverrides
      };

      return shallow(<JobCard {...props} />);
    };

    it('expect name to be rendered', () => {
      const component = renderComponent();
      const field = component.find('.job-card__name');
      expect(field.text()).toMatch('Jean-Ralphio Saperstein');
      expect(component).toMatchSnapshot();
    });

    it('expect start time to be rendered and formatted', () => {
      const component = renderComponent();
      const field = component.find('.job-card__start-time');
      expect(field.text()).toMatch(moment(defaultProps.start).format(DATE_FORMAT));
      expect(component).toMatchSnapshot();
    });

    it('expect end time to be rendered and formatted', () => {
      const component = renderComponent();
      const field = component.find('.job-card__end-time');
      expect(field.text()).toMatch(moment(defaultProps.end).format(DATE_FORMAT));
      expect(component).toMatchSnapshot();
    });

    it('expect contact name to be rendered', () => {
      const component = renderComponent();
      const field = component.find('.job-card__contact-name');
      expect(field.text()).toMatch('Tom Haverford');
      expect(component).toMatchSnapshot();
    });

  });

});