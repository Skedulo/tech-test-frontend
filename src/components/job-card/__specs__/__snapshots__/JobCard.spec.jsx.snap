// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobCard Component snapshots expect contact name to be rendered 1`] = `
<div
  className="job-card__container"
>
  <p
    className="job-card__name"
  >
    <span>
      Job
    </span>
    Jean-Ralphio Saperstein
  </p>
  <p
    className="job-card__start-time"
  >
    <span>
      Start Time
    </span>
    September 2nd 2018, 1:00:00 am
  </p>
  <p
    className="job-card__end-time"
  >
    <span>
      End Time
    </span>
    September 2nd 2018, 2:00:00 am
  </p>
  <p
    className="job-card__contact-name"
  >
    <span>
      Contact
    </span>
    Tom Haverford
  </p>
</div>
`;

exports[`JobCard Component snapshots expect end time to be rendered and formatted 1`] = `
<div
  className="job-card__container"
>
  <p
    className="job-card__name"
  >
    <span>
      Job
    </span>
    Jean-Ralphio Saperstein
  </p>
  <p
    className="job-card__start-time"
  >
    <span>
      Start Time
    </span>
    September 2nd 2018, 1:00:00 am
  </p>
  <p
    className="job-card__end-time"
  >
    <span>
      End Time
    </span>
    September 2nd 2018, 2:00:00 am
  </p>
  <p
    className="job-card__contact-name"
  >
    <span>
      Contact
    </span>
    Tom Haverford
  </p>
</div>
`;

exports[`JobCard Component snapshots expect name to be rendered 1`] = `
<div
  className="job-card__container"
>
  <p
    className="job-card__name"
  >
    <span>
      Job
    </span>
    Jean-Ralphio Saperstein
  </p>
  <p
    className="job-card__start-time"
  >
    <span>
      Start Time
    </span>
    September 2nd 2018, 1:00:00 am
  </p>
  <p
    className="job-card__end-time"
  >
    <span>
      End Time
    </span>
    September 2nd 2018, 2:00:00 am
  </p>
  <p
    className="job-card__contact-name"
  >
    <span>
      Contact
    </span>
    Tom Haverford
  </p>
</div>
`;

exports[`JobCard Component snapshots expect start time to be rendered and formatted 1`] = `
<div
  className="job-card__container"
>
  <p
    className="job-card__name"
  >
    <span>
      Job
    </span>
    Jean-Ralphio Saperstein
  </p>
  <p
    className="job-card__start-time"
  >
    <span>
      Start Time
    </span>
    September 2nd 2018, 1:00:00 am
  </p>
  <p
    className="job-card__end-time"
  >
    <span>
      End Time
    </span>
    September 2nd 2018, 2:00:00 am
  </p>
  <p
    className="job-card__contact-name"
  >
    <span>
      Contact
    </span>
    Tom Haverford
  </p>
</div>
`;
